src/elements/core/testing/private/visual-regression-snapshot.ts

```diff
@@ -20,61 +22,121 @@ function findElementCenter(snapshotElement: () => HTMLElement): [number, number]
-      await visualDiff(snapshotElement(), imageName(this.test!));
-    } finally {
-      await resetMouse();
-    }
-  });
+  public async withFixture(
+    template: TemplateResult,
+    wrapperStyles?: Parameters<typeof visualRegressionFixture>[1],
+  ): Promise<this> {
+    this._snapshotElement = await visualRegressionFixture(template, wrapperStyles);
+    return this;
+  }
 }
 
-export function visualRegressionSnapshot(
-  snapshotElement: () => HTMLElement,
-  stateElement?: () => HTMLElement,
-): void {
-  testVisualDiff(snapshotElement);
-  testVisualDiffFocus(snapshotElement);
-  testVisualDiffHover(snapshotElement, stateElement);
-  testVisualDiffActive(snapshotElement, stateElement);
+const applyViewportIfDefined = async (
+  viewport: { width: number; height: number } | undefined,
+): Promise<void> => {
+  if (viewport) {
+    await setViewport(viewport);
+  }
+};
+
+export interface VisualDiffState {
+  name: string;
+  with: (setup: (setup: VisualDiffSetup) => void | Promise<void>) => Mocha.Func;
 }
+
+export const visualDiffDefault: VisualDiffState = {
+  name: 'default',
+  with(setup: (setup: VisualDiffSetup) => void | Promise<void>): Mocha.Func {
+    return async function (this: Mocha.Context) {
+      const context = new VisualDiffSetup();
+      await setup(context);
+      await applyViewportIfDefined(this.ctx['requestViewport']);
+      await visualDiff(context.snapshotElement, imageName(this.test!));
+    };
+  },
+};
+
+export const visualDiffFocus: VisualDiffState = {
+  name: 'default',
+  with(setup: (setup: VisualDiffSetup) => void | Promise<void>): Mocha.Func {
+    return async function (this: Mocha.Context) {
+      const context = new VisualDiffSetup();
+      await setup(context);
+      await applyViewportIfDefined(this.ctx['requestViewport']);
+      context.snapshotElement.focus();
+      await sendKeys({ press: 'Tab' });
+      await visualDiff(context.snapshotElement, imageName(this.test!));
+    };
+  },
+};
+
+export const visualDiffHover: VisualDiffState = {
+  name: 'default',
+  with(setup: (setup: VisualDiffSetup) => void | Promise<void>): Mocha.Func {
+    return async function (this: Mocha.Context) {
+      const context = new VisualDiffSetup();
+      await setup(context);
+      await applyViewportIfDefined(this.ctx['requestViewport']);
+      const position = findElementCenter(context.stateElement);
+      try {
+        await sendMouse({ type: 'move', position });
+        await aTimeout(5);
+        await visualDiff(context.snapshotElement, imageName(this.test!));
+      } finally {
+        await resetMouse();
+      }
+    };
+  },
+};
+
+export const visualDiffActive: VisualDiffState = {
+  name: 'default',
+  with(setup: (setup: VisualDiffSetup) => void | Promise<void>): Mocha.Func {
+    return async function (this: Mocha.Context) {
+      const context = new VisualDiffSetup();
+      await setup(context);
+      await applyViewportIfDefined(this.ctx['requestViewport']);
```

Comment:
Those three lines are appearing multiple times. shouldn't we create a method or include it more globally?